### PR TITLE
Lifespan QoS [4992]

### PR DESCRIPTION
**Pull request by raquelalvarezbanos**
Implements lifespan QoS. Adds blackbox tests and C++ example.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1356/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1319/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1333/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1358/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1320/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1334/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1364/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1328/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1342/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1376/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1344/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1356/)

File: examples/C++/LifespanQoSExample/LifespanPublisher.cpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file LifespanPublisher.cpp
+ *
+ */
+
+#include "LifespanPublisher.h"
+#include <fastrtps/participant/Participant.h>
```
Comment (richiware) [inline examples/C++/LifespanQoSExample/LifespanPublisher.cpp]: Lifespan example is new. Can you use our coding style on it?

File: examples/C++/LifespanQoSExample/README.txt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+This example illustrates how to use the Lifespan QoS.
+
+To launch this test open two different consoles:
```
Comment (richiware) [inline examples/C++/LifespanQoSExample/README.txt]: Can you rewrite in markdown format?

File: examples/C++/LifespanQoSExample/LifespanPublisher.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    Domain::registerType(mp_participant,&m_type);
+
+    PublisherAttributes Wparam;
+    Wparam.topic.topicKind = NO_KEY;
+    Wparam.topic.topicDataType = "Lifespan";
+    Wparam.topic.topicName = "LifespanTopic";
+    Wparam.topic.historyQos.kind = KEEP_ALL_HISTORY_QOS;
+    Wparam.topic.historyQos.depth = 30;
+    Wparam.qos.m_reliability.kind = RELIABLE_RELIABILITY_QOS;
```
Comment (richiware) [inline examples/C++/LifespanQoSExample/LifespanPublisher.cpp]: Even it is the default value, I will add the QoS durability to `TRANSIENT_LOCAL`. So the user could figure out that is the reason the example can remove samples from history after being sent.

File: include/fastrtps/participant/Participant.h [hunk 4]
```diff
@@ -86,6 +87,11 @@ class RTPS_DllAPI Participant
     bool get_remote_writer_info(const rtps::GUID_t& writerGuid, rtps::WriterProxyData& returnedInfo);
 
     bool get_remote_reader_info(const rtps::GUID_t& readerGuid, rtps::ReaderProxyData& returnedInfo);
+
+    /**
+     * Get the resource event for this participant
+     */
+    rtps::ResourceEvent& get_resource_event() const;
```
Comment (richiware) [inline include/fastrtps/participant/Participant.h]: Why is it needed at public API level?

File: include/fastrtps/rtps/participant/RTPSParticipant.h [hunk 5]
```diff
@@ -141,6 +141,8 @@ class RTPS_DllAPI RTPSParticipant
 
     bool get_remote_reader_info(const GUID_t& readerGuid, ReaderProxyData& returnedInfo);
 
+    ResourceEvent& get_resource_event() const;
```
Comment (richiware) [inline include/fastrtps/rtps/participant/RTPSParticipant.h]: Why this is needed at Public API level?

File: include/fastrtps/rtps/resources/TimedCallback.h [hunk 6]
```diff
@@ -0,0 +1,67 @@
+// Copyright 2019 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
```
Comment (richiware) [inline include/fastrtps/rtps/resources/TimedCallback.h]: I think a best location for this file would be `include/fastrtps/timedevent`

File: examples/C++/LifespanQoSExample/LifespanPublisher.cpp (hunk 1 above)
Comment (raquelalvarezbanos) [reply examples/C++/LifespanQoSExample/LifespanPublisher.cpp]: :+1: Changed

File: examples/C++/LifespanQoSExample/README.txt (hunk 2 above)
Comment (raquelalvarezbanos) [reply examples/C++/LifespanQoSExample/README.txt]: :+1:

File: examples/C++/LifespanQoSExample/LifespanPublisher.cpp (hunk 3 above)
Comment (raquelalvarezbanos) [reply examples/C++/LifespanQoSExample/LifespanPublisher.cpp]: Added

File: include/fastrtps/participant/Participant.h (hunk 4 above)
Comment (raquelalvarezbanos) [reply include/fastrtps/participant/Participant.h]: As in the deadline case, I must have thought it was needed at some point but I've removed it now.

File: include/fastrtps/rtps/resources/TimedCallback.h (hunk 6 above)
Comment (raquelalvarezbanos) [reply include/fastrtps/rtps/resources/TimedCallback.h]: Moved

File: include/fastrtps/rtps/participant/RTPSParticipant.h (hunk 5 above)
Comment (raquelalvarezbanos) [reply include/fastrtps/rtps/participant/RTPSParticipant.h]: This one I believe is needed: 
- to initialize the `TimedCallback` I need to get the resource event, which is owned by `RTPSParticipantImpl`
- but `TimedCallback` lives in `PublisherImpl`, which only has access to `ParticipantImpl`

So I needed to get the resource like this:

```
ParticipantImpl::get_resource_event()
    -> RTPSParticipant::get_resource_event()
        -> RTPSParticipantImpl::get_resource_event()
            -> return resource_event;
```

Is there an alternative? This is the same I did in the Deadline QoS

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1382/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1350/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1362/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1389/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1357/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1369/)

File: include/fastrtps/participant/Participant.h [hunk 7]
```diff
@@ -35,6 +35,7 @@ namespace rtps
 {
     class WriterProxyData;
     class ReaderProxyData;
+    class ResourceEvent;
```
Comment (richiware) [inline include/fastrtps/participant/Participant.h]: This is already not necessary.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1409/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1378/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1389/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1498/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1463/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1474/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1509/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1474/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1485/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1519/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1483/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1494/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1523/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1486/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1497/)

**Pull request merged**